File: src/main/scala/com/yanli/flink/streaming/kafka/FlinkConnectKafka.scala
package com.yanli.flink.streaming.kafka



import java.lang
import java.nio.charset.StandardCharsets

import com.yanli.flink.config.KafkaConfig
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.streaming.connectors.kafka.{FlinkKafkaConsumer, FlinkKafkaProducer, KafkaSerializationSchema}
import org.apache.flink.api.scala._
import org.apache.kafka.clients.producer.ProducerRecord


/**
 * @author yanli
 * @date 2019/12/22 16:13
 * @version 1.0
 */
object FlinkConnectKafka {
  def getKafkaSource(env: StreamExecutionEnvironment,topic: String): DataStream[String] ={
    val flinkConsumer: FlinkKafkaConsumer[String] = new FlinkKafkaConsumer[String](topic, new SimpleStringSchema(), KafkaConfig.getKafkaConsumerConfig())
    flinkConsumer.setStartFromLatest()    //设置开始消费的位置为latest

    val data: DataStream[String] = env.addSource(flinkConsumer)

    data
  }

  def getKafkaSink(dataStream :DataStream[String],topic :String): Unit ={
    val flinkProducer: FlinkKafkaProducer[String] = new FlinkKafkaProducer[String](
      topic,
      //新的api需要用KafkaSerializationSchema ,跟SimpleStringSchema 同样的实现
      new KafkaSerializationSchema[String] {
        override def serialize(element: String, timestamp: lang.Long): ProducerRecord[Array[Byte], Array[Byte]] = new ProducerRecord[Array[Byte], Array[Byte]](topic,element.getBytes(StandardCharsets.UTF_8))
      },
      KafkaConfig.getKafkaProducerConfig(),
      FlinkKafkaProducer.Semantic.EXACTLY_ONCE
    )

    flinkProducer.setWriteTimestampToKafka(true)

    dataStream.addSink(flinkProducer)
  }

}

File: src/main/java/com/yanli/flink/java/streamingApi/kafka/FlinkConnectKafka.java
package com.yanli.flink.java.streamingApi.kafka;

import com.yanli.flink.java.config.KafkaConfig;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer;
import org.apache.flink.streaming.connectors.kafka.KafkaSerializationSchema;
import org.apache.kafka.clients.producer.ProducerRecord;

import javax.annotation.Nullable;
import java.nio.charset.StandardCharsets;

/**
 * @author YanLi
 * @version 1.0
 * @ClassName: FlinkConnectKafka
 * @date 2021/2/24 5:11 下午
 */
public class FlinkConnectKafka {
    /**
     *  创建kafka source
     * @param environment
     * @param topic
     * @return
     */
    public static DataStream getKafkaSource(StreamExecutionEnvironment environment, String topic,String bootstrap_servers) {
        FlinkKafkaConsumer<String> flinkKafkaConsumer = new FlinkKafkaConsumer<String>(topic, new SimpleStringSchema(), KafkaConfig.getKafkaConsumerConfig(bootstrap_servers));
        flinkKafkaConsumer.setStartFromLatest();
        flinkKafkaConsumer.setCommitOffsetsOnCheckpoints(true);

        return environment.addSource(flinkKafkaConsumer);
    }

    /**
     * 创建kafka sink
     * @param dataStream
     * @param topic
     * @param topicKey
     */
    public static void getKafkaSink(DataStream dataStream, String topic, String topicKey) {
        FlinkKafkaProducer<String> flinkKafkaProducer = new FlinkKafkaProducer<String>(topic, new KafkaSerializationSchema<String>() {
            @Override
            public ProducerRecord<byte[], byte[]> serialize(String element, @Nullable Long timestamp) {
                return new ProducerRecord(topic, topicKey.getBytes(StandardCharsets.UTF_8), element.getBytes(StandardCharsets.UTF_8));
            }
        }, KafkaConfig.getKafkaProducerConfig()
                , FlinkKafkaProducer.Semantic.EXACTLY_ONCE);

        flinkKafkaProducer.setWriteTimestampToKafka(true);

        dataStream.addSink(flinkKafkaProducer);
    }
}
